core/store/src/utils/shard_size.rs

```diff
@@ -0,0 +1,146 @@
+            get(state_update, &trie_key)?.unwrap_or_default();
+        account_contribution += contribution;
+        set(state_update, trie_key, &account_contribution);
+    }
+    Ok(())
+}
+
+/// Find an account ID that splits the shard into two possibly equal parts, according
+/// to the shard size contribution.
+/// Returns `None` if the trie is empty or an empty node is found.
+pub fn find_split_account(trie: &Trie) -> Result<Option<AccountId>, StorageError> {
+    // Find the split key
+    let key = match trie.read_lock_memtries() {
+        Some(memtries) => find_split_key(MemTrieIteratorInner::new(&*memtries, trie))?,
+        None => find_split_key(DiskTrieIteratorInner::new(trie))?,
+    };
+    let Some(key) = key else {
+        return Ok(None);
+    };
+
+    // Convert the key to account ID by finding the longest prefix that is a valid ID
+    for i in 0..key.len() {
+        let prefix = std::str::from_utf8(&key[..(key.len() - i)]).expect("trie key is non-ASCII");
+        if let Ok(account_id) = AccountId::from_str(prefix) {
+            return Ok(Some(account_id));
+        }
+    }
+    Ok(None)
+}
+
+fn find_split_key<NodePtr, ValueRef, TrieStorage>(
+    trie_storage: TrieStorage,
+) -> Result<Option<Vec<u8>>, StorageError>
+where
+    NodePtr: Copy,
+    TrieStorage: GenericTrieInternalStorage<NodePtr, ValueRef>,
+{
+    let mut key_nibbles = Vec::with_capacity(AccountId::MAX_LEN * 2); // x2 because nibbles are half-bytes
+    let Some(mut current_node_ptr) = trie_storage.get_root() else {
+        return Ok(None);
+    };
+    loop {
+        match trie_storage.get_node(current_node_ptr, AccessOptions::DEFAULT)? {
+            GenericTrieNode::Empty => return Ok(None),
+            GenericTrieNode::Leaf { extension, .. } => {
+                key_nibbles.extend(extension);
+                return Ok(Some(NibbleSlice::nibbles_to_bytes(&key_nibbles)));
+            }
+            GenericTrieNode::Extension { extension, child } => {
+                key_nibbles.extend(extension);
+                current_node_ptr = child;
+            }
+            GenericTrieNode::Branch { children, value } => {
+                let Some(parent_value_ref) = value else { return Ok(None) };
+                let parent_contribution = get_value_contribution(&trie_storage, parent_value_ref)?;
+                let Some((idx, child)) =
+                    find_middle_child(&trie_storage, parent_contribution, *children)?
+                else {
+                    return Ok(Some(NibbleSlice::nibbles_to_bytes(&key_nibbles)));
+                };
+                key_nibbles.push(idx);
+                current_node_ptr = child;
+            }
+        }
+    }
+}
+
+fn find_middle_child<NodePtr, ValueRef, TrieStorage>(
+    trie_storage: &TrieStorage,
+    parent_contribution: ShardSizeContribution,
+    children: [Option<NodePtr>; 16],
+) -> Result<Option<(u8, NodePtr)>, StorageError>
```

Comment:
I think you also need context from the entire path to the root about how much is on the left and on the right from the current path.

Reply:
Also it may happen that the best solution is actually outside of the current subtree -> e.g.
```
       16
   /    |   \   
 3      5     8
       / \
      3   2
```
your current algorithm will return 16, 5, 3
adjusting for context it would return 16, 5, 2
where the best split is actually 16, 8

Reply:
I think to find 16,8 in this case we just need to replace `>=` with `>` (this mistake is due to inclusive/exclusive boundary confusion). But the context is required indeed. We should be adding everything on the right and comparing it against the total size stored in the root.